students/DennisLee/lesson07/test_html_render.py

```diff
@@ -0,0 +1,691 @@
+#!/usr/bin/env python3
+
+import unittest, os
+import html_render as hr
+
+class consts():
+    """Constant strings/dicts to be used in during testing."""
+    strs_before = (
+            "  \n Why  am   \n\tI here?     \t",
+            "\n\t  \n  Here's   that rainy   day...   \t  \n",
+            "I told  you   about...\n",
+            "    \n \t  \n  There's  a   somebody I'm longing  to   see\n",
+            "   \n \t \t Someone  who'll \twatch\nover\nme   \t   ",
+            "\tI\tdid\nit\t\tmy      \n\n\t\tway\t\t\n\t"
+    )
+    strs_after = (
+            "Why am I here?", 
+            "Here's that rainy day...",
+            "I told you about...",
+            "There's a somebody I'm longing to see",
+            "Someone who'll watch over me",
+            "I did it my way"
+    )
+    attrs_fail = {
+        'id': 60,
+        'alt': '  \t  \n \tFancy      \t\t   Pants   \n\n\t\n '
+    }
+    attrs_pass = {
+        'id': 'Marker',
+        'alt': '  \t  \n \tFancy      \t\t   Pants   \n\n\t\n '
+    }
+    random_tag = " \t SomeTagName \n "
+    test_filename = "test_html_file_dennislee.html"
+
+
+class ElementTestCase(unittest.TestCase):
+    def setUp(self):
+        """
+        Initialize output string list, stream/file objects, and three
+        test elements. Also make sure the test input and output string
+        lists have the same number of items.
+        """
+        self.strs_out, self.file, self.fobj = None, None, None
+        self.e1, self.e2, self.e3 = None, None, None
+        self.assertEqual(len(consts.strs_before), len(consts.strs_after))        
+    def tearDown(self):
+        """Delete setup objects and test file (if necessary)."""
+        del self.strs_out, self.file, self.fobj
+        del self.e1, self.e2, self.e3
```

Comment:
what is the purpose of above?

Reply:
I thought we needed to clean up objects manually from memory after we're done with them?

I was experiencing lots of problems while using iPython to test code. I even discovered that if I ran tests in iPython, closed and then reopened iPython without changing any of the implementation or test code, and reran the tests again, I would get differing test results. At that point I decided to manually delete objects to avoid any data persistence issues and to run the tests using python.exe from the normal command line (instead of iPython). This was one of a few reasons why this assignment took me so long to complete.

Reply:
No you don't need to do that. There had to be something else going on causing side effects in the tests.